File: 39-combination-sum/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
// Backtracking
// O(n * (n^t))...n for the # of nums and t for target
var combinationSum = function(nums, target) {
    const result = [], n = nums.length

    const backtrack = (idx, curr, t) => {
        if ( idx >= n || t < 0) return;
        if (t === 0) return result.push([...curr])

        for (let i = idx; i < n; i++) {
            curr.push(nums[i])
            backtrack(i, curr, t - nums[i])
            curr.pop()
        }
    }
    backtrack(0, [], target)
    return result;
};

File: 39-combination-sum/combination-sum.swift
class Solution {
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        var result: [[Int]] = []

        func recurse(idx: Int, combo: inout [Int], sum: Int) -> [Int] {
            if sum == 0 {
                result.append(combo.map{$0})
                return combo
            }
            if idx >= candidates.count { return combo }
            if sum < 0 { return combo }

            for i in idx..<candidates.count {
                if sum - candidates[i] >= 0 {
                    combo.append(candidates[i])
                    recurse(idx: i, combo: &combo, sum: sum - candidates[i])
                    combo.popLast()
                }
            }
            return combo
        }
        var temp: [Int] = []
        recurse(idx: 0, combo: &temp, sum: target)
        return result
    }
}